src/db/car/many.rs

```diff
@@ -49,40 +88,53 @@ impl<WriterT> ManyCar<WriterT> {
     pub fn with_read_only<ReaderT: super::RandomAccessFileReader>(
         self,
         any_car: AnyCar<ReaderT>,
-    ) -> Self {
-        self.read_only(any_car);
-        self
+    ) -> anyhow::Result<Self> {
+        self.read_only(any_car)?;
+        Ok(self)
     }
 
-    pub fn read_only<ReaderT: super::RandomAccessFileReader>(&self, any_car: AnyCar<ReaderT>) {
+    fn read_only<ReaderT: super::RandomAccessFileReader>(
+        &self,
+        any_car: AnyCar<ReaderT>,
+    ) -> anyhow::Result<()> {
         let mut read_only = self.read_only.write();
         let key = read_only.len() as u64;
-        read_only.push(
-            any_car
-                .with_cache(self.shared_cache.clone(), key)
-                .into_dyn(),
-        );
+
+        let car = any_car
+            .with_cache(self.shared_cache.clone(), key)
+            .into_dyn();
+        read_only
+            .push(WithHeaviestEpoch::new(car).context("store doesn't have a heaviest tipset")?);
+
+        Ok(())
     }
 
-    pub fn with_read_only_files(self, files: impl Iterator<Item = PathBuf>) -> io::Result<Self> {
+    pub fn with_read_only_files(
+        self,
+        files: impl Iterator<Item = PathBuf>,
+    ) -> anyhow::Result<Self> {
         self.read_only_files(files)?;
         Ok(self)
     }
 
-    pub fn read_only_files(&self, files: impl Iterator<Item = PathBuf>) -> io::Result<()> {
+    pub fn read_only_files(&self, files: impl Iterator<Item = PathBuf>) -> anyhow::Result<()> {
         for file in files {
-            self.read_only(AnyCar::new(EitherMmapOrRandomAccessFile::open(file)?)?);
+            let car = AnyCar::new(EitherMmapOrRandomAccessFile::open(&file)?)?;
+            debug!("Loaded car DB at {}", file.display());
+
+            self.read_only(car)?;
         }
 
         Ok(())
     }
 
+    // TODO: update
     pub fn heaviest_tipset(&self) -> anyhow::Result<Tipset> {
         let tipsets = self
```

Comment:
Should it be sufficient to load heaviest from the first CAR now?

Reply:
Yes, I'm using `peek` method.